modulemd_tools/modulemd_tools/yaml.py

```diff
@@ -0,0 +1,317 @@
+
+    supported = [1, 2]
+    if parsed["version"] not in supported or version not in supported:
+        raise ValueError("Unexpected modulemd version")
+
+    if parsed["version"] > version:
+        raise ValueError("Cannot downgrade modulemd version")
+
+    mod_stream = Modulemd.ModuleStream.new(parsed["version"])
+    mod_stream = mod_stream.read_string(mod_yaml, True,
+                                        parsed["data"].get("name", ""),
+                                        parsed["data"].get("stream", ""))
+    mod_stream_upgraded = mod_stream.upgrade(version)
+    return _stream2yaml(mod_stream_upgraded)
+
+
+def load(path):
+    """
+    Load modulemd YAML definition from a file
+    """
+    with open(path, "r") as yaml_file:
+        mod_yaml = yaml_file.read()
+    validate(mod_yaml)
+    return mod_yaml
+
+
+def dump(mod_yaml, dest=None):
+    """
+    Dump modulemd YAML definition into a file
+
+    If `dest` is not specified, the file will be created in the current working
+    directory and it's name is going to be generated from the module attributes
+    in the `N:S:V:C:A.modulemd.yaml` format.
+
+    If `dest` is a directory, then the file is going to be generated in that
+    directory.
+
+    If `dest` is pointing to a (nonexisting) file, it is going to be dumped in
+    its place.
+    """
+    filename = _generate_filename(mod_yaml)
+
+    path = dest
+    if not path:
+        path = os.path.join(os.getcwd(), filename)
+    elif os.path.isdir(path):
+        path = os.path.join(dest, filename)
+
+    with open(path, "w") as moduleyaml:
+        moduleyaml.write(mod_yaml)
+
+
+def _generate_filename(mod_yaml):
+    """
+    Generate filename for a module yaml
+    """
+    mod_stream = _yaml2stream(mod_yaml)
+    return "{N}:{S}:{V}:{C}:{A}.modulemd.yaml".format(
+        N=mod_stream.get_module_name(),
+        S=mod_stream.get_stream_name(),
+        V=mod_stream.get_version(),
+        C=mod_stream.get_context(),
+        A=mod_stream.get_arch() or "noarch")
+
+
+def _yaml2stream(mod_yaml):
+    if not mod_yaml:
+        raise ValueError("Provided YAML did not begin with a module document")
+
+    idx = Modulemd.ModuleIndex.new()
+    idx.update_from_string(mod_yaml, strict=True)
+
+    mod_names = idx.get_module_names()
+    if len(mod_names) > 1:
+        raise ValueError("Provided YAML contains multiple modules {0}. "
+                           "It is ambiguous which one to use."
+                           .format(mod_names))
+
+    mod_name = idx.get_module_names()[0]
+    mod = idx.get_module(mod_name)
+    streams = mod.get_all_streams()
+    if len(streams) > 1:
+        raise ValueError("Provided YAML contains multiple streams {0} of "
+                           "a module '{1}'. It is ambiguous which one to use."
+                           .format([x.get_stream_name() for x in streams],
+                                   mod_name))
+    return streams[0]
```

Comment:
This entire function is unnecessary. You can use
```
Modulemd.ModuleStream.read_string(mod_yaml, strict=True|False)
```

This returns a `Modulemd.ModuleStream` object rather than a `Modulemd.ModuleIndex` and throws an exception if the document contains anything other than a single Stream subdocument.